File: src/make_oracle.py
"""Génère l'oracle de validation pour le port C++.

Lance le pipeline sur les N premières frames de la vidéo et sauve les sorties
bit-exactes (boxes, polygones, scores, textes OCR, hash des masques) en JSON.
Le port C++ devra reproduire ces valeurs à l'identique (cf. ROADMAP_CPP.md).

Usage : python -m src.make_oracle [--video PATH] [--frames N] [--out PATH]
"""
import argparse
import hashlib
import json
import os

import cv2

from . import config
from .models.yolo import load_yolo
from .models.ocr import load_ocr
from .pipeline.core import make_initial_state, run_pipeline

_PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


def _default_video() -> str:
    if os.path.exists(config.INPUT_VIDEO):
        return config.INPUT_VIDEO
    return os.path.join(_PROJECT_ROOT, "rush_2.avi")


def _serialize_detection(det: dict) -> dict:
    mask = det.get("mask")
    return {
        "class": str(det["class"]),
        "score": float(det["score"]),
        "box": [int(v) for v in det["box"]],
        "polygon": [[float(x), float(y)] for x, y in det["polygon"]],
        "text": det.get("text", ""),
        "mask_shape": list(mask.shape) if mask is not None else None,
        "mask_sha256": hashlib.sha256(mask.tobytes()).hexdigest() if mask is not None else None,
    }


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--video", default=_default_video())
    parser.add_argument("--frames", type=int, default=40)
    parser.add_argument("--out", default=os.path.join(_PROJECT_ROOT, "oracle", "oracle_40frames.json"))
    args = parser.parse_args()

    cap = cv2.VideoCapture(args.video)
    if not cap.isOpened():
        raise SystemExit(f"Impossible d'ouvrir la vidéo : {args.video}")

    yolo = load_yolo(config.YOLO_MODEL, config.PROVIDERS)
    ocr = load_ocr(config.OCR_MODEL, config.CHARS_PATH, config.PROVIDERS)
    state = make_initial_state()

    frames = []
    for i in range(args.frames):
        ok, frame = cap.read()
        if not ok:
            print(f"Vidéo épuisée à la frame {i}")
            break
        _, detections, state, _ = run_pipeline(frame, state, yolo, ocr)
        frames.append({
            "frame": i,
            "detections": [_serialize_detection(d) for d in detections],
        })
    cap.release()

    oracle = {
        "video": os.path.basename(args.video),
        "n_frames": len(frames),
        "conf": config.CONF,
        "iou": config.IOU,
        "ocr_frame_interval": config.OCR_FRAME_INTERVAL,
        "providers": yolo.session.get_providers(),
        "frames": frames,
    }
    os.makedirs(os.path.dirname(args.out), exist_ok=True)
    with open(args.out, "w", encoding="utf-8") as f:
        json.dump(oracle, f, indent=1)

    n_det = sum(len(fr["detections"]) for fr in frames)
    print(f"Oracle écrit : {args.out} ({len(frames)} frames, {n_det} détections)")


if __name__ == "__main__":
    main()

File: cpp/src/make_oracle.cpp
// Miroir de src/make_oracle.py — dump JSON des sorties pipeline sur N frames,
// à comparer à oracle/oracle_40frames.json via tools/compare_oracle.py.
// Usage : lapi_oracle <video> <yolo.onnx> <ocr.onnx> <en_dict.txt> <out.json> [n_frames]
#include <fstream>
#include <iomanip>
#include <iostream>
#include <limits>

#include "config.hpp"
#include "io/reader.hpp"
#include "models/ocr.hpp"
#include "models/yolo.hpp"
#include "pipeline/core.hpp"

namespace {

void write_detection(std::ostream& os, const lapi::Detection& det) {
    os << "{\"class\": \"" << det.klass << "\", \"score\": " << det.score
       << ", \"box\": [" << det.box.x << ", " << det.box.y << ", "
       << det.box.x + det.box.width << ", " << det.box.y + det.box.height
       << "], \"polygon\": [";
    for (int i = 0; i < 4; ++i) {
        if (i) os << ", ";
        os << "[" << det.polygon[i].x << ", " << det.polygon[i].y << "]";
    }
    os << "], \"text\": \"" << det.text << "\", \"mask_shape\": ["
       << det.mask.rows << ", " << det.mask.cols << "]}";
}

}  // namespace

int main(int argc, char** argv) {
    using namespace lapi;
    if (argc < 6) {
        std::cerr << "Usage : lapi_oracle <video> <yolo.onnx> <ocr.onnx> "
                     "<en_dict.txt> <out.json> [n_frames=40]\n";
        return 2;
    }
    const std::string video_path = argv[1];
    const int n_frames = argc > 6 ? std::stoi(argv[6]) : 40;

    try {
        Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "lapi");
        YoloModel yolo = load_yolo(argv[2], env);
        OcrModel ocr = load_ocr(argv[3], argv[4], env);

        cv::VideoCapture cap = open_video(video_path);
        PipelineState state = make_initial_state();

        std::ofstream out(argv[5]);
        out << std::setprecision(std::numeric_limits<double>::max_digits10);
        out << "{\"video\": \"" << video_path << "\", \"n_frames\": " << n_frames
            << ", \"conf\": " << config::CONF << ", \"iou\": " << config::IOU
            << ", \"ocr_frame_interval\": " << config::OCR_FRAME_INTERVAL
            << ", \"frames\": [";

        cv::Mat frame;
        int n = 0;
        int total_det = 0;
        for (; n < n_frames && cap.read(frame); ++n) {
            PipelineResult res = run_pipeline(frame, state, yolo, ocr);
            if (n) out << ", ";
            out << "{\"frame\": " << n << ", \"detections\": [";
            for (size_t i = 0; i < res.detections.size(); ++i) {
                if (i) out << ", ";
                write_detection(out, res.detections[i]);
            }
            out << "]}";
            total_det += static_cast<int>(res.detections.size());
        }
        out << "]}\n";
        std::cout << "Oracle C++ écrit : " << argv[5] << " (" << n
                  << " frames, " << total_det << " détections)\n";
    } catch (const std::exception& e) {
        std::cerr << "Erreur : " << e.what() << "\n";
        return 1;
    }
    return 0;
}
